src/main/java/io/harness/cf/client/connector/HarnessConnector.java

```diff
@@ -0,0 +1,187 @@
+              // comment with code
+              Response response = chain.proceed(request);
+              if (response.code() == 403 && onUnauthorized != null) {
+                onUnauthorized.run();
+              }
+              return response;
+            })
+        .addInterceptor(new RetryInterceptor(3, 2000));
+    return apiClient;
+  }
+
+  protected ApiClient makeMetricsApiClient() {
+    final int maxTimeout = 30 * 60 * 1000;
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(options.getEventUrl());
+    apiClient.setConnectTimeout(maxTimeout);
+    apiClient.setReadTimeout(maxTimeout);
+    apiClient.setWriteTimeout(maxTimeout);
+    apiClient.setDebugging(log.isDebugEnabled());
+    apiClient.getHttpClient().newBuilder().addInterceptor(new RetryInterceptor(3, 2000));
+    return apiClient;
+  }
+
+  @Override
+  public Optional<String> authenticate(Consumer<String> onError) {
+    try {
+      final AuthenticationRequest request = AuthenticationRequest.builder().apiKey(apiKey).build();
+      final AuthenticationResponse response = api.authenticate(request);
+      token = response.getAuthToken();
+      processToken(token);
+      return Optional.of(token);
+    } catch (ApiException apiException) {
+      log.error("Failed to get auth token {}", apiException.getMessage());
+
+      if (apiException.getCode() == 401 || apiException.getCode() == 403) {
+
+        String errorMsg = String.format("Invalid apiKey %s. Serving default value. ", apiKey);
+
+        log.error(errorMsg);
+        onError.accept(errorMsg);
+      }
+
+      onError.accept(apiException.getMessage());
+    }
+    return Optional.empty();
+  }
+
+  protected void processToken(@NonNull String token) {
+    api.getApiClient().addDefaultHeader("Authorization", String.format("Bearer %s", token));
+    metricsApi.getApiClient().addDefaultHeader("Authorization", "Bearer " + token);
+
+    // get claims
+    int i = token.lastIndexOf('.');
+    String unsignedJwt = token.substring(0, i + 1);
+    Jwt<?, Claims> untrusted = Jwts.parserBuilder().build().parseClaimsJwt(unsignedJwt);
+
+    environment = (String) untrusted.getBody().get("environment");
+    cluster = (String) untrusted.getBody().get("clusterIdentifier");
+  }
+
+  @Override
+  public List<FeatureConfig> getFlags() {
+    try {
+      return api.getFeatureConfig(environment, cluster);
+    } catch (ApiException e) {
+      e.printStackTrace();
+      return null;
+    }
+  }
+
+  @Override
+  public Optional<FeatureConfig> getFlag(@NonNull String identifier) {
+    try {
+      return Optional.of(api.getFeatureConfigByIdentifier(identifier, environment, cluster));
+    } catch (ApiException e) {
+      e.printStackTrace();
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public List<Segment> getSegments() {
+    try {
+      return api.getAllSegments(environment, cluster);
+    } catch (ApiException e) {
+      e.printStackTrace();
+      return null;
```

Comment:
Consider returning empty list instead of null on all occurances. Usually that's better instead of null when result is a collection